### PR TITLE
docs/usage: Add build and run complex apps doc

### PR DESCRIPTION
**Pull request by mariasfiraiala**
This PR adds documentation for building and running complex applications on `x86` and `AArch64`.

It goes into detail regarding configuration steps and manual launch commands for the make build system.
Depends on having these PRs merged:

* [core Unikraft virtio PR](https://github.com/unikraft/unikraft/pull/519)
* [`newlib` PR](https://github.com/unikraft/lib-newlib/pull/21)

Signed-off-by: Maria Sfiraiala <maria.sfiraiala@gmail.com>

### CODE REVIEW

**Comment by mariasfiraiala** [comment]
With the transition to `musl` finally stable, I got the time to revisit this PR and finalize the documentation. @StefanJum @razvand please take a look when you have the time.

File: content/en/docs/usage/make_build.md [hunk 1]
```diff
@@ -461,3 +461,453 @@ oOo oOO| | | | |   (| | | (_) |  _) :_
...
+
+Now, onto configuring our app:
+
+```console
+$ make menuconfig
+```
+
+1. choose our usuals from `Architecture Selection`
+
+1. choose all `Virtio` options from `Platform configuration`->`KVM guest`->`Virtio`
```
[the remark this answers is not in the document]
Comment (StefanJum) [reply content/en/docs/usage/make_build.md:521]: Shouldn't `musl` also be here?

File: content/en/docs/usage/make_build.md [hunk 2]
```diff
@@ -461,3 +461,453 @@ oOo oOO| | | | |   (| | | (_) |  _) :_
...
+And testing it:
+
+```console
+Booting from ROM..Powered by
+o.   .o       _ _               __ _
+Oo   Oo  ___ (_) | __ __  __ _ ' _) :_
+oO   oO ' _ `| | |/ /  _)' _` | |_|  _)
+oOo oOO| | | | |   (| | | (_) |  _) :_
+ OoOoO ._, ._:_:_,\_._,  .__,_:_, \___)
+                  Phoebe 0.10.0~9bf6e63
```
Comment (StefanJum) [inline content/en/docs/usage/make_build.md:569]: Change this to the `0.12` unikraft version :rocket:

**Review by StefanJum — CHANGES_REQUESTED** [review]
Thanks a lot @mariasfiraiala, see my comments.

**Comment by mariasfiraiala** [comment]
Done @StefanJum! Take another tour when you find the time.

File: content/en/docs/usage/make_build.md [hunk 3]
```diff
@@ -461,3 +461,455 @@ oOo oOO| | | | |   (| | | (_) |  _) :_
                  Phoebe 0.10.0~ee2a37cf
 Hello world!
 ```
+
+## Building and Running Complex Applications with Make
+
+Let's say we decide to use already ported applications.
+There are many to choose from, so you go for `app-sqlite`, `app-redis` or `app-nginx`.
```
Comment (StefanJum) [inline content/en/docs/usage/make_build.md:468]:
```suggestion
There are many to choose from, so we will go for `app-sqlite`, `app-redis` or `app-nginx`.
```

File: content/en/docs/usage/make_build.md [hunk 4]
```diff
@@ -461,3 +461,455 @@ oOo oOO| | | | |   (| | | (_) |  _) :_
...
+This tutorial is here to guide you through the first steps towards successfully launching your (more complex) unikernel; we'll present 2 scenarios: building for `x86` and `AArch64`.
```
Comment (StefanJum) [inline content/en/docs/usage/make_build.md:469]:
```suggestion
This tutorial is here to guide us through the first steps towards successfully launching our (more complex) unikernel; we'll present 2 scenarios: building for `x86` and `AArch64`.
```

File: content/en/docs/usage/make_build.md [hunk 5]
```diff
@@ -461,3 +461,455 @@ oOo oOO| | | | |   (| | | (_) |  _) :_
...
+-2505304097900087778|-1087908404908377458
+sqlite> .exit
+```
+
+**Pro tip**: We usually save the launch command in a script.
+This way, it's easier to type it, but if you fancy even more automation, check out [`qemu-guest`](https://github.com/unikraft/kraft/blob/staging/scripts/qemu-guest) or even [kraft](https://unikraft.org/docs/usage/init/#kraft-run).
+
+### `app-sqlite` for `AArch64`
+
+The process of building `app-sqlite` for `AArch64` is quite similar to the one we've layed out for the `x86` architecture, the only thing that needs changes being the architecture, which should be switched to `Armv8` in the `Architecture Selection` menu.
```
Comment (StefanJum) [inline content/en/docs/usage/make_build.md:596]: Maybe also specify that a properclean must be done after changing the architecture.

File: content/en/docs/usage/make_build.md [hunk 6]
```diff
@@ -461,3 +461,455 @@ oOo oOO| | | | |   (| | | (_) |  _) :_
...
+LIBS := $(UK_LIBS)/lib-musl:$(UK_LIBS)/lib-lwip:$(UK_LIBS)/lib-redis
+
+all:
+	@$(MAKE) -C $(UK_ROOT) A=$(PWD) L=$(LIBS)
+
+$(MAKECMDGOALS):
+	@$(MAKE) -C $(UK_ROOT) A=$(PWD) L=$(LIBS) $(MAKECMDGOALS)
+```
+
+We should alos create a `Makefile.uk` in the `app-redis` directory:
```
Comment (StefanJum) [inline content/en/docs/usage/make_build.md:646]:
```suggestion
We should also create a `Makefile.uk` in the `app-redis` directory:
```

File: content/en/docs/usage/make_build.md [hunk 7]
```diff
@@ -461,3 +461,455 @@ oOo oOO| | | | |   (| | | (_) |  _) :_
...
+                           -netdev bridge,id=en0,br=kraft0 \
+                           -device virtio-net-pci,netdev=en0 \
+                           -kernel "build/app-nginx_kvm-arm64" \
+                           -append "netdev.ipv4_addr=172.44.0.2 netdev.ipv4_gw_addr=172.44.0.1 netdev.ipv4_subnet_mask=255.255.255.0 --" \
+                           -machine virt \
+                           -cpu cortex-a57 \
+                           -nographic
+```
+
+With this last command, our tutorial ends here, it was a long ride but the results should be rewarding enough.
```
Comment (StefanJum) [inline content/en/docs/usage/make_build.md:915]:
```suggestion
With this last command, our tutorial ends, it was a long ride but the results should be rewarding enough.
```

File: content/en/docs/usage/make_build.md [hunk 8]
```diff
@@ -461,3 +461,455 @@ oOo oOO| | | | |   (| | | (_) |  _) :_
...
+1. Choose all `Virtio` options from `Platform configuration->KVM guest->Virtio`
+
+1. Choose the `SQLite`, `musl` and `vfscore` libraries from the `Library Configuration` screen.
+
+1. Under the `vfscore` library, choose the `Default root filesystem` to be `9pfs`
+
+We'll build our app:
+
+```console
+$ make
```
Comment (StefanJum) [inline content/en/docs/usage/make_build.md:532]:
```suggestion
$ make -j $(ncpus)
```

File: content/en/docs/usage/make_build.md [hunk 9]
```diff
@@ -461,3 +461,455 @@ oOo oOO| | | | |   (| | | (_) |  _) :_
...
+1. In the `Redis` library choose the `Provide main function` option
+
+1. In the `lwip` library make sure to have `IPv4`, `UDP support`, `TCP support`, `ICMP support`, `DHCP support`, `Socket API` enabled
+
+1. Under the `vfscore` library, choose the `Default root filesystem` to be `9pfs`
+
+We'll build our app:
+
+```console
+$ make
```
Comment (StefanJum) [inline content/en/docs/usage/make_build.md:676]:
```suggestion
$ make -j $(ncpus)
```

File: content/en/docs/usage/make_build.md [hunk 10]
```diff
@@ -461,3 +461,455 @@ oOo oOO| | | | |   (| | | (_) |  _) :_
...
+1. In the `libnginx` library, choose the `Provide main function` option
+
+1. In the `lwip` library, make sure to have `IPv4`, `UDP support`, `TCP support`, `ICMP support`, `DHCP support`, `Socket API` enabled
+
+1. Under the `vfscore` library, choose the `Default root filesystem` to be `9pfs`
+
+We'll build our app:
+
+```console
+$ make
```
Comment (StefanJum) [inline content/en/docs/usage/make_build.md:812]:
```suggestion
$ make -j $(ncpus)
```

**Review by StefanJum — CHANGES_REQUESTED** [review]
Looks really nice @mariasfiraiala, thank you.
I will test the command in the next few days and come back with a new round.

**Comment by mariasfiraiala** [comment]
Great thing @StefanJum. I've also integrated your latest suggestions. Hope it looks alright now (Markdown wise).

File: content/en/docs/usage/make_build.md [hunk 11]
```diff
@@ -461,3 +461,461 @@ oOo oOO| | | | |   (| | | (_) |  _) :_
...
+
+```console
+$ make menuconfig
+```
+
+1. Choose the `x86` architecture from `Architecture Selection`
+
+1. Choose the `KVM` platform from `Platform Configuration`
+
+1. Choose the `Redis`, `lwip`, `musl` libraries from `Library Configuration`
```
Comment (StefanJum) [inline content/en/docs/usage/make_build.md:671]:
```suggestion
1. Choose the `Redis`, `lwip` and `musl` libraries from `Library Configuration`
```

File: content/en/docs/usage/make_build.md [hunk 12]
```diff
@@ -461,3 +461,461 @@ oOo oOO| | | | |   (| | | (_) |  _) :_
...
+
+```console
+$ sudo ip l set dev kraft0 down
+$ sudo brctl delbr kraft0
+```
+
+### `app-redis` for `AArch64`
+
+Building `Redis` for the `AArch64` architecture looks very similar to how we did it for `x86`, the only thing that needs changes being the architecture, which should be switched to `Armv8` in the `Architecture Selection` menu.
+
```
Comment (StefanJum) [inline content/en/docs/usage/make_build.md:739]: As on the `sqlite` app, mention that a `properclean` is needed here.

File: content/en/docs/usage/make_build.md [hunk 13]
```diff
@@ -461,3 +461,461 @@ oOo oOO| | | | |   (| | | (_) |  _) :_
...
+|      |---lib-nginx/
+|---unikraft/
+```
+
+We must create a `Makefile` in the `app-nginx` directory:
+
+```Makefile
+UK_ROOT ?= $(PWD)/../../unikraft
+UK_LIBS ?= $(PWD)/../../libs
+LIBS := $(UK_LIBS)/lib-musl:$(UK_LIBS)/lib-lwip:$(UK_LIBS)/lib-redis
```
Comment (StefanJum) [inline content/en/docs/usage/make_build.md:782]:
```suggestion
LIBS := $(UK_LIBS)/lib-musl:$(UK_LIBS)/lib-lwip:$(UK_LIBS)/lib-nginx
```

File: content/en/docs/usage/make_build.md [hunk 14]
```diff
@@ -461,3 +461,461 @@ oOo oOO| | | | |   (| | | (_) |  _) :_
...
+
+```console
+$ sudo ip l set dev kraft0 down
+$ sudo brctl delbr kraft0
+```
+
+### `app-nginx` for `AArch64`
+
+Just like we've stated for `SQLite` and `Redis`, for `AArch64` the build steps follow the tutorial for `x86`, the only thing that needs changes being the architecture, which should be switched to `Armv8` in the `Architecture Selection` menu.
+
```
Comment (StefanJum) [inline content/en/docs/usage/make_build.md:906]: As on the `sqlite` app, mention that a `properclean` is needed here.

**Review by StefanJum — CHANGES_REQUESTED** [review]
Great work @mariasfiraiala, I tested the commands and everything works well.
See the few comments I left.

**Comment by mariasfiraiala** [comment]
I've committed your suggestions @StefanJum. Take another look!

File: content/en/docs/usage/make_build.md [hunk 15]
```diff
@@ -461,3 +461,482 @@ oOo oOO| | | | |   (| | | (_) |  _) :_
...
+The first step is to clone our dependencies:
+
+1. [unikraft](https://github.com/unikraft/unikraft)
+
+1. [app-sqlite](https://github.com/unikraft/app-sqlite)
+
+1. [lib-sqlite](https://github.com/unikraft/lib-sqlite), [lib-musl](https://github.com/unikraft/lib-musl)
+
+We should be left with a file hierarchy that should look like this:
```
Comment (StefanJum) [inline content/en/docs/usage/make_build.md:482]:
```suggestion
We should be left with a file hierarchy that looks like this:
```

File: content/en/docs/usage/make_build.md [hunk 16]
```diff
@@ -461,3 +461,482 @@ oOo oOO| | | | |   (| | | (_) |  _) :_
...
+
+The first step is to clone our dependencies:
+
+1. [unikraft](https://github.com/unikraft/unikraft)
+
+1. [app-redis](https://github.com/unikraft/app-redis)
+
+1. [lib-redis](https://github.com/unikraft/lib-redis), [lib-musl](https://github.com/unikraft/lib-musl), [lib-lwip](https://github.com/unikraft/lib-lwip)
+
+We should be left with a file hierarchy that should look like this:
```
Comment (StefanJum) [inline content/en/docs/usage/make_build.md:630]:
```suggestion
We should be left with a file hierarchy that looks like this:
```

File: content/en/docs/usage/make_build.md [hunk 17]
```diff
@@ -461,3 +461,482 @@ oOo oOO| | | | |   (| | | (_) |  _) :_
...
+
+The first step is to clone our dependencies:
+
+1. [unikraft](https://github.com/unikraft/unikraft)
+
+1. [app-nginx](https://github.com/unikraft/app-nginx)
+
+1. [lib-nginx](https://github.com/unikraft/lib-nginx), [lib-musl](https://github.com/unikraft/lib-musl), [lib-lwip](https://github.com/unikraft/lib-lwip)
+
+We should be left with a file hierarchy that should look like this:
```
Comment (StefanJum) [inline content/en/docs/usage/make_build.md:777]:
```suggestion
We should be left with a file hierarchy that looks like this:
```

**Review by StefanJum — CHANGES_REQUESTED** [review]
All good besides the few comments, thanks @mariasfiraiala.
@razvand please take a quick look over this.

**Review by StefanJum — APPROVED** [review]
All good, thanks.

Reviewed-by: Stefan Jumarea <stefanjumarea02@gmail.com>
Approved-by: Stefan Jumarea <stefanjumarea02@gmail.com>

File: content/en/docs/usage/make_build.md [hunk 18]
```diff
@@ -461,3 +461,482 @@ oOo oOO| | | | |   (| | | (_) |  _) :_
                  Phoebe 0.10.0~ee2a37cf
 Hello world!
 ```
+
+## Building and Running Complex Applications with Make
+
+Let's say we decide to use already ported applications.
+There are many to choose from, so we will go for [`app-sqlite`](#app-sqlite-for-x86), [`app-redis`](#app-redis-for-x86) or [`app-nginx`](#app-nginx-for-x86).
+This tutorial is here to guide us through the first steps towards successfully launching our (more complex) unikernel.
+For every application, we'll present 2 scenarios: building for `x86` and `AArch64`.
```
Comment (razvand) [inline content/en/docs/usage/make_build.md:470]:
```suggestion
For every application, we'll present 2 scenarios: building for `x86_64` and `AArch64`.
```

File: content/en/docs/usage/make_build.md [hunk 19]
```diff
@@ -461,3 +461,482 @@ oOo oOO| | | | |   (| | | (_) |  _) :_
...
+
+### `app-sqlite` for `x86`
```
Comment (razvand) [inline content/en/docs/usage/make_build.md:472]:
```suggestion
### `app-sqlite` for `x86_64`
```

File: content/en/docs/usage/make_build.md [hunk 20]
```diff
@@ -461,3 +461,482 @@ oOo oOO| | | | |   (| | | (_) |  _) :_
...
+-861992803967016915|1583603794332302036
+-5556545375765846451|-8514616268392191775
+6518576960059048853|6826969221610647599
+-2505304097900087778|-1087908404908377458
+sqlite> .exit
+```
+
+{{% alert theme="info" %}}
+**Pro tip**: We usually save the launch command in a script.
+This way, it's easier to type it, but if you fancy even more automation, check out [`qemu-guest`](https://github.com/unikraft/kraft/blob/staging/scripts/qemu-guest) or even [`kraft`](https://unikraft.org/docs/usage/init/#kraft-run).
```
Comment (razvand) [inline content/en/docs/usage/make_build.md:594]:
```suggestion
This way, it's easier to type it, but if you fancy even more automation, check out [`qemu-guest`](https://github.com/unikraft/kraft/blob/staging/scripts/qemu-guest) or [`kraft`](https://unikraft.org/docs/usage/init/#kraft-run).
```

File: content/en/docs/usage/make_build.md [hunk 21]
```diff
@@ -461,3 +461,482 @@ oOo oOO| | | | |   (| | | (_) |  _) :_
...
+{{% /alert %}}
+
+### `app-sqlite` for `AArch64`
+
+The process of building `app-sqlite` for `AArch64` is quite similar to the one we've layed out for the `x86` architecture, the only thing that needs changes being the architecture, which should be switched to `Armv8` in the `Architecture Selection` menu.
```
[the remark this answers is not in the document]
Comment (razvand) [reply content/en/docs/usage/make_build.md:599]: Remind that configuration happens with `make menuconfig`.

File: content/en/docs/usage/make_build.md [hunk 22]
```diff
@@ -461,3 +461,482 @@ oOo oOO| | | | |   (| | | (_) |  _) :_
...
+
+1. Choose the `KVM` platform from `Platform Configuration`
+
+1. Choose the `Redis`, `lwip` and `musl` libraries from `Library Configuration`
+
+1. In the `Redis` library choose the `Provide main function` option
+
+1. In the `lwip` library make sure to have `IPv4`, `UDP support`, `TCP support`, `ICMP support`, `DHCP support`, `Socket API` enabled
+
+1. Under the `vfscore` library, choose the `Default root filesystem` to be `9pfs`
```
Comment (razvand) [inline content/en/docs/usage/make_build.md:682]:
```suggestion
1. Under the `vfscore` library, choose the `Default root filesystem` to be `9pfs`.
```

File: content/en/docs/usage/make_build.md [hunk 23]
```diff
@@ -461,3 +461,482 @@ oOo oOO| | | | |   (| | | (_) |  _) :_
...
+$(eval $(call addlib,appnginx))
+```
+
+Onto configuring our app:
+
+```console
+$ make menuconfig
+```
+
+1. Choose the `x86` architecture from `Architecture Selection`
```
Comment (razvand) [inline content/en/docs/usage/make_build.md:816]: Add dots at the end.

File: content/en/docs/usage/make_build.md [hunk 24]
```diff
@@ -461,3 +461,482 @@ oOo oOO| | | | |   (| | | (_) |  _) :_
...
+qemu-system-x86_64: bridge helper failed
+```
+
+make sure to add this to your `/etc/qemu/bridge.conf` file (if it doesn't exist, create it):
+
+```console
+allow all
+```
+
+For testing our app we'll open a new terminal, or use `tmux`, and give this command to retrive data from the server
```
Comment (razvand) [inline content/en/docs/usage/make_build.md:881]:
```suggestion
For testing our app, we'll open a new terminal, or use `tmux`, and give this command to retrive data from the server:
```

File: content/en/docs/usage/make_build.md [hunk 25]
```diff
@@ -461,3 +461,482 @@ oOo oOO| | | | |   (| | | (_) |  _) :_
...
+Don't forget to clean up our work: let's delete the bridge we've created earlier:
+
+```console
+$ sudo ip l set dev kraft0 down
+$ sudo brctl delbr kraft0
+```
+
+### `app-nginx` for `AArch64`
+
+Just like we've stated for `SQLite` and `Redis`, for `AArch64` the build steps follow the tutorial for `x86`, the only thing that needs changes being the architecture, which should be switched to `Armv8` in the `Architecture Selection` menu.
```
Comment (razvand) [inline content/en/docs/usage/make_build.md:918]:
```suggestion
Just like we've stated for `SQLite` and `Redis`, for `AArch64` the build steps follow the tutorial for `x86_64`.
The only thing that needs changes being the architecture, which should be switched to `Armv8` in the `Architecture Selection` menu.
```

**Review by razvand — CHANGES_REQUESTED** [review]
Hi, @mariasfiraiala. This looks very good. See my minor comments.

I want to take another tour, to test the steps you mentioned and make sure everything is clear for future readers.

**Comment by mariasfiraiala** [comment]
Done @razvand. Take another look, please.

**Review by razvand — CHANGES_REQUESTED** [review]
Hi, @mariasfiraiala, see my comments. Also use `virbr0` as the bridge ID, instead of `kraft0`, throughout the tutorial.

File: content/en/docs/usage/make_build.md [hunk 26]
```diff
@@ -461,3 +461,485 @@ oOo oOO| | | | |   (| | | (_) |  _) :_
...
+$(MAKECMDGOALS):
+	@$(MAKE) -C $(UK_ROOT) A=$(PWD) L=$(LIBS) $(MAKECMDGOALS)
+```
+
+We should also create a `Makefile.uk` in the `app-nginx` directory:
+
+```Makefile
+$(eval $(call addlib,appnginx))
+```
+
```
[the remark this answers is not in the document]
Comment (mariasfiraiala) [reply content/en/docs/usage/make_build.md:811]: Not really, since we already provide it in the [app repo](https://github.com/unikraft/app-nginx).

**Comment by mariasfiraiala** [comment]
Just integrated what you've suggested @razvand! Thanks for the time and take another look, please!

**Review by razvand — APPROVED** [review]
Thanks, @mariasfiraiala.

Reviewed-by: Razvan Deaconescu <razvand@unikraft.io>
Approved-by: Razvan Deaconescu <razvand@unikraft.io>

**Pull request merged**